pkg/volume/csi/csi_mounter.go

```diff
@@ -107,49 +111,97 @@ func (c *csiMountMgr) SetUpAt(dir string, fsGroup *int64) error {
 
-	if stageUnstageSet {
-		deviceMountPath, err = makeDeviceMountPath(c.plugin, c.spec)
-		if err != nil {
-			klog.Error(log("mounter.SetUpAt failed to make device mount path: %v", err))
-			return err
+	driverName := c.driverName
+	volumeHandle := c.volumeID
+	readOnly := c.readOnly
+	accessMode := api.ReadWriteOnce
+
+	var (
+		fsType             string
+		volAttribs         map[string]string
+		nodePublishSecrets map[string]string
+		publishContext     map[string]string
+		mountOptions       []string
+		deviceMountPath    string
+		secretRef          *api.SecretReference
+	)
+
+	switch {
+	case volSrc != nil:
+		if !utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
+			return fmt.Errorf("CSIInlineVolume feature required")
 		}
-	}
-	// search for attachment by VolumeAttachment.Spec.Source.PersistentVolumeName
-	if c.publishContext == nil {
-		nodeName := string(c.plugin.host.GetNodeName())
-		c.publishContext, err = c.plugin.getPublishContext(c.k8s, c.volumeID, string(c.driverName), nodeName)
+		if c.driverMode != ephemeralDriverMode {
+			return fmt.Errorf("unexpected driver mode: %s", c.driverMode)
+		}
+		if volSrc.FSType != nil {
+			fsType = *volSrc.FSType
+		}
+
+		volAttribs = volSrc.VolumeAttributes
+
+		if volSrc.NodePublishSecretRef != nil {
+			secretName := volSrc.NodePublishSecretRef.Name
+			ns := c.pod.Namespace
+			secretRef = &api.SecretReference{Name: secretName, Namespace: ns}
+		}
+	case pvSrc != nil:
+		if c.driverMode != persistentDriverMode {
+			return fmt.Errorf("unexpected driver mode: %s", c.driverMode)
+		}
+
+		fsType = pvSrc.FSType
+
+		volAttribs = pvSrc.VolumeAttributes
+
+		if pvSrc.NodePublishSecretRef != nil {
+			secretRef = pvSrc.NodePublishSecretRef
+		}
+
+		//TODO (vladimirvivien) implement better AccessModes mapping between k8s and CSI
+		if c.spec.PersistentVolume.Spec.AccessModes != nil {
+			accessMode = c.spec.PersistentVolume.Spec.AccessModes[0]
+		}
+
+		mountOptions = c.spec.PersistentVolume.Spec.MountOptions
+
+		// Check for STAGE_UNSTAGE_VOLUME set and populate deviceMountPath if so
+		stageUnstageSet, err := csi.NodeSupportsStageUnstage(ctx)
 		if err != nil {
+			klog.Error(log("mounter.SetUpAt failed to check for STAGE_UNSTAGE_VOLUME capabilty: %v", err))
 			return err
 		}
-	}
 
-	attribs := csiSource.VolumeAttributes
+		if stageUnstageSet {
+			deviceMountPath, err = makeDeviceMountPath(c.plugin, c.spec)
+			if err != nil {
+				klog.Error(log("mounter.SetUpAt failed to make device mount path: %v", err))
+				return err
+			}
+		}
 
-	nodePublishSecrets := map[string]string{}
-	if csiSource.NodePublishSecretRef != nil {
-		nodePublishSecrets, err = getCredentialsFromSecret(c.k8s, csiSource.NodePublishSecretRef)
-		if err != nil {
-			return fmt.Errorf("fetching NodePublishSecretRef %s/%s failed: %v",
-				csiSource.NodePublishSecretRef.Namespace, csiSource.NodePublishSecretRef.Name, err)
+		// search for attachment by VolumeAttachment.Spec.Source.PersistentVolumeName
```

Comment:
what does this comment mean? it doesn't seem to have anything to do with surrounding calls

Reply:
Will check.